dj/models/history.py

```diff
@@ -0,0 +1,61 @@
+"""
+Model for history.
+"""
+from datetime import datetime, timezone
+from enum import Enum
+from functools import partial
+from typing import Any, Dict, Optional
+
+from sqlalchemy import DateTime
+from sqlalchemy.sql.schema import Column as SqlaColumn
+from sqlmodel import JSON, Field, SQLModel
+
+from dj.typing import UTCDatetime
+
+
+class ActivityType(str, Enum):
+    """
+    An activity type
+    """
+
+    CREATE = "create"
+    DEACTIVATE = "deactivate"
+    ACTIVATE = "activate"
+    UPDATE = "modify"
+    LINK = "link"
+    TAG = "tag"
+    SET_ATTRIBUTE = "set_attribute"
+
+
+class EntityType(str, Enum):
+    """
+    An entity type for which activity can occur
+    """
+
+    ATTRIBUTE = "attribute"
+    CATALOG = "catalog"
+    ENGINE = "engine"
+    NAMESPACE = "namespace"
+    NODE = "node"
+    QUERY = "query"
+    TAG = "tag"
+
+
+class History(SQLModel, table=True):  # type: ignore
+    """
+    An event to store as part of the server's activity history
+    """
+
+    id: Optional[int] = Field(default=None, primary_key=True)
+    entity_type: Optional[EntityType] = Field(default=None)
+    entity_name: Optional[str] = Field(default=None)
+    activity_type: Optional[ActivityType] = Field(default=None)
+    user: Optional[str] = Field(default=None)
+    details: Dict[str, Any] = Field(default={}, sa_column=SqlaColumn(JSON))
```

Comment:
Wondering how we can represent the `details` for each type of activity. I think it should be easy for most event types, but what about UPDATE?

Reply:
What if we added a `pre: Dict[str, Any]` and `post: Dict[str, Any]` attribute and these can be populated when the history is captured? How the diff is displayed can then be handled client said depending on the environment and context. For example the UI may use something similar to [react-diff-viewer](https://github.com/praneshr/react-diff-viewer/tree/v2.0).

Reply:
Added the pre and post attributes.